File: pipeline/reimpl/cbstm03b.py
"""
Reimplementation of CBSTM03B — CardDemo Statement File I/O Subroutine.

CBSTM03A calls CBSTM03B as a subroutine passing an operation code:
  'O' = Open file
  'C' = Close file
  'R' = Read next record
  'K' = Read by key
  'W' = Write record
  'Z' = Rewrite record

This subroutine manages the transaction file and returns records/status codes.
"""

from __future__ import annotations
from dataclasses import dataclass

from .carddemo_data import TranRecord


# ── Operation codes (88-level condition names in COBOL) ───────────────────────

OPER_OPEN    = "O"
OPER_CLOSE   = "C"
OPER_READ    = "R"
OPER_READ_K  = "K"
OPER_WRITE   = "W"
OPER_REWRITE = "Z"


@dataclass
class M03BArea:
    """WS-M03B-AREA — linkage area between CBSTM03A and CBSTM03B."""
    dd: str = "TRNXFILE"   # dataset name
    oper: str = ""          # operation code
    rc: str = "00"          # return code: '00'=ok, '10'=EOF, '23'=not-found
    key: str = ""           # record key for keyed reads
    key_ln: int = 0         # length of key
    fldt: str = ""          # record data area (1000 chars)


class Cbstm03bFileManager:
    """Stateful file I/O manager, injected with a transaction list."""

    def __init__(self, transactions: list[TranRecord]):
        self._transactions = transactions
        self._cursor = 0
        self._by_id: dict[str, TranRecord] = {t.tran_id: t for t in transactions}
        self._is_open = False

    def execute(self, area: M03BArea) -> M03BArea:
        """Process a single CALL 'CBSTM03B' invocation."""
        op = area.oper

        if op == OPER_OPEN:
            self._cursor = 0
            self._is_open = True
            area.rc = "00"

        elif op == OPER_CLOSE:
            self._is_open = False
            area.rc = "00"

        elif op == OPER_READ:
            if not self._is_open:
                area.rc = "35"
            elif self._cursor >= len(self._transactions):
                area.rc = "10"  # EOF
            else:
                tran = self._transactions[self._cursor]
                self._cursor += 1
                area.fldt = self._serialize(tran)
                area.rc = "00"

        elif op == OPER_READ_K:
            key = area.key.strip()
            tran = self._by_id.get(key)
            if tran is None:
                area.rc = "23"  # not found
            else:
                area.fldt = self._serialize(tran)
                area.rc = "00"

        elif op == OPER_WRITE:
            tran = self._deserialize(area.fldt)
            if tran.tran_id not in self._by_id:
                self._transactions.append(tran)
                self._by_id[tran.tran_id] = tran
                area.rc = "00"
            else:
                area.rc = "22"  # duplicate key

        elif op == OPER_REWRITE:
            tran = self._deserialize(area.fldt)
            if tran.tran_id in self._by_id:
                self._by_id[tran.tran_id] = tran
                area.rc = "00"
            else:
                area.rc = "23"  # not found

        else:
            area.rc = "99"

        return area

    def _serialize(self, tran: TranRecord) -> str:
        """Serialize a TranRecord into the 1000-char FLDT area."""
        return (
            f"{tran.tran_id:<16}"
            f"{tran.tran_type_cd:<2}"
            f"{tran.tran_cat_cd:04d}"
            f"{tran.tran_source:<10}"
            f"{tran.tran_desc:<100}"
            f"{float(tran.tran_amt):014.2f}"
            f"{tran.tran_merchant_id:09d}"
            f"{tran.tran_merchant_name:<50}"
            f"{tran.tran_merchant_city:<50}"
            f"{tran.tran_merchant_zip:<10}"
            f"{tran.tran_card_num:<16}"
            f"{tran.tran_orig_ts:<26}"
            f"{tran.tran_proc_ts:<26}"
        )

    def _deserialize(self, fldt: str) -> TranRecord:
        """Deserialize FLDT area back into a TranRecord (best-effort)."""
        from decimal import Decimal
        try:
            return TranRecord(
                tran_id=fldt[0:16].strip(),
                tran_type_cd=fldt[16:18].strip(),
                tran_cat_cd=int(fldt[18:22] or 0),
                tran_source=fldt[22:32].strip(),
                tran_desc=fldt[32:132].strip(),
                tran_amt=Decimal(fldt[132:146].strip() or "0"),
                tran_merchant_id=int(fldt[146:155] or 0),
                tran_merchant_name=fldt[155:205].strip(),
                tran_merchant_city=fldt[205:255].strip(),
                tran_merchant_zip=fldt[255:265].strip(),
                tran_card_num=fldt[265:281].strip(),
                tran_orig_ts=fldt[281:307].strip(),
                tran_proc_ts=fldt[307:333].strip(),
            )
        except Exception:
            return TranRecord()

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cbstm03b.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Java reimplementation of CardDemo subroutine CBSTM03B — Statement File I/O.
 *
 * <p>Mirrors {@code pipeline/reimpl/cbstm03b.py} byte-for-byte. Python is the
 * source of truth.
 */
public class Cbstm03b implements ProgramRunner {

    private static final String OPER_OPEN  = "O";
    private static final String OPER_CLOSE = "C";
    private static final String OPER_READ  = "R";

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenarioName = inputs.getOrDefault("SCENARIO", "PROCESS_RECORDS").toUpperCase();
        List<TranRecord> trans = buildScenario(scenarioName);
        if (trans == null) {
            Map<String, String> err = new LinkedHashMap<>();
            err.put("error", "unknown scenario: '" + scenarioName + "'");
            return err;
        }

        // Create file manager
        FileManager mgr = new FileManager(trans);

        // OPEN
        M03BArea area = new M03BArea();
        area.oper = OPER_OPEN;
        mgr.execute(area);
        String openRc = area.rc;

        // READ all
        int recordsRead = 0;
        List<String> fldtRecords = new ArrayList<>();
        while (true) {
            area.oper = OPER_READ;
            mgr.execute(area);
            if ("10".equals(area.rc)) {
                break;
            }
            recordsRead++;
            fldtRecords.add(area.fldt);
        }

        // CLOSE
        area.oper = OPER_CLOSE;
        mgr.execute(area);
        String closeRc = area.rc;

        Map<String, String> out = new LinkedHashMap<>();
        out.put("OPEN_RC", openRc);
        out.put("CLOSE_RC", closeRc);
        out.put("RECORDS_READ", String.valueOf(recordsRead));
        for (int i = 0; i < fldtRecords.size(); i++) {
            out.put("FLDT_" + i, fldtRecords.get(i));
        }
        return out;
    }

    // ── Scenarios ─────────────────────────────────────────────────────────

    private List<TranRecord> buildScenario(String name) {
        switch (name) {
            case "PROCESS_RECORDS": return scenarioProcessRecords();
            case "EMPTY_INPUT":    return scenarioEmptyInput();
            default: return null;
        }
    }

    private List<TranRecord> scenarioProcessRecords() {
        List<TranRecord> list = new ArrayList<>();
        list.add(new TranRecord(
                "TRN0000000000001", "01", 1,
                "ONLINE", "Widget purchase",
                new BigDecimal("150.25"), 123456789,
                "ACME WIDGETS", "NEW YORK", "10001",
                "4111000000001111",
                "2026-04-01-12.00.00.000000",
                "2026-04-01-12.00.00.000000"
        ));
        return list;
    }

    private List<TranRecord> scenarioEmptyInput() {
        return new ArrayList<>();
    }

    // ── File Manager (mirrors Cbstm03bFileManager) ───────────────────────

    static class M03BArea {
        String dd = "TRNXFILE";
        String oper = "";
        String rc = "00";
        String key = "";
        int keyLn = 0;
        String fldt = "";
    }

    static class FileManager {
        private final List<TranRecord> transactions;
        private int cursor = 0;
        private boolean isOpen = false;

        FileManager(List<TranRecord> transactions) {
            this.transactions = transactions;
        }

        void execute(M03BArea area) {
            String op = area.oper;
            if (OPER_OPEN.equals(op)) {
                cursor = 0;
                isOpen = true;
                area.rc = "00";
            } else if (OPER_CLOSE.equals(op)) {
                isOpen = false;
                area.rc = "00";
            } else if (OPER_READ.equals(op)) {
                if (!isOpen) {
                    area.rc = "35";
                } else if (cursor >= transactions.size()) {
                    area.rc = "10";
                } else {
                    TranRecord tran = transactions.get(cursor);
                    cursor++;
                    area.fldt = serialize(tran);
                    area.rc = "00";
                }
            } else {
                area.rc = "99";
            }
        }

        private String serialize(TranRecord tran) {
            return String.format("%-16s%-2s%04d%-10s%-100s%014.2f%09d%-50s%-50s%-10s%-16s%-26s%-26s",
                    tran.tranId, tran.tranTypeCd, tran.tranCatCd,
                    tran.tranSource, tran.tranDesc,
                    tran.tranAmt.doubleValue(), tran.tranMerchantId,
                    tran.tranMerchantName, tran.tranMerchantCity,
                    tran.tranMerchantZip, tran.tranCardNum,
                    tran.tranOrigTs, tran.tranProcTs);
        }
    }

    // ── Inner data types ─────────────────────────────────────────────────

    static class TranRecord {
        final String tranId;
        final String tranTypeCd;
        final int tranCatCd;
        final String tranSource;
        final String tranDesc;
        final BigDecimal tranAmt;
        final int tranMerchantId;
        final String tranMerchantName;
        final String tranMerchantCity;
        final String tranMerchantZip;
        final String tranCardNum;
        final String tranOrigTs;
        final String tranProcTs;

        TranRecord(String tranId, String tranTypeCd, int tranCatCd,
                   String tranSource, String tranDesc, BigDecimal tranAmt,
                   int tranMerchantId, String tranMerchantName,
                   String tranMerchantCity, String tranMerchantZip,
                   String tranCardNum, String tranOrigTs, String tranProcTs) {
            this.tranId = tranId;
            this.tranTypeCd = tranTypeCd;
            this.tranCatCd = tranCatCd;
            this.tranSource = tranSource;
            this.tranDesc = tranDesc;
            this.tranAmt = tranAmt;
            this.tranMerchantId = tranMerchantId;
            this.tranMerchantName = tranMerchantName;
            this.tranMerchantCity = tranMerchantCity;
            this.tranMerchantZip = tranMerchantZip;
            this.tranCardNum = tranCardNum;
            this.tranOrigTs = tranOrigTs;
            this.tranProcTs = tranProcTs;
        }
    }
}
